packages/protocol/contracts-0.8/common/EpochManager.sol

```diff
@@ -494,4 +555,73 @@ contract EpochManager is
       CELOequivalent
     );
   }
+
+  /**
+   * @notice Returns the epoch info of a specified blockNumber.
+   * @param _blockNumber Block number of the epoch info is retreived.
+   * @return firstEpoch The first block of the given block number.
+   * @return lastBlock The first block of the given block number.
+   * @return startTimestamp The starting timestamp of the given block number.
+   * @return rewardsBlock The reward block of the given block number.
+   * @return elected The set of elected validator for the given block number.
+   */
+  function _getEpochByBlockNumber(
+    uint256 _blockNumber
+  )
+    internal
+    view
+    onlySystemAlreadyInitialized
+    returns (uint256, uint256, uint256, uint256, uint256, address[] memory)
+  {
+    require(_blockNumber <= block.number, "Invalid blockNumber. Value too high.");
+
+    (uint256 _firstBlockOfFirstEpoch, , , , ) = getEpochByNumber(firstKnownEpoch);
+
+    require(_blockNumber >= _firstBlockOfFirstEpoch, "Invalid blockNumber. Value too low.");
+
+    uint256 _firstBlockOfCurrentEpoch = epochs[currentEpochNumber].firstBlock;
+
+    if (_blockNumber > _firstBlockOfCurrentEpoch) {
+      (
+        uint256 _firstBlock,
+        uint256 _lastBlock,
+        uint256 _startTimestamp,
+        uint256 _rewardsBlock,
+        address[] memory _elected
+      ) = getEpochByNumber(currentEpochNumber);
+      return (
+        currentEpochNumber,
+        _firstBlock,
+        _lastBlock,
+        _startTimestamp,
+        _rewardsBlock,
+        _elected
+      );
+    }
+
+    uint256 left = firstKnownEpoch;
+    uint256 right = currentEpochNumber - 1;
+
+    while (left <= right) {
+      uint256 mid = (left + right) / 2;
+      Epoch memory _epoch = epochs[mid];
```

Comment:
Would this be much more gas efficient with only grabbing a `storage` ref (or reading only `first`/`lastBlock` into memory for the binsearch)? Not sure if the compiler optimizes away reading the other fields (including the entire `elected` array) into memory, which could incur a lot of unnecessary `SLOAD`s while searching.

Reply:
This struct no longer includes the election history